transform/snowflake-dbt/models/events/user_events_telemetry_precalc.sql

```diff
@@ -0,0 +1,220 @@
+{{config({
+    "materialized": "incremental",
+    "schema": "events",
+    "tags":"nightly",
+    "snowflake_warehouse": "transform_l",
+    "cluster_by": ['event_date']
+  })
+}}
+
+-- User events telemetry table, but with precalculated fields. This is an attempt to temporarily improve the
+-- performance of user_events_telemetry table. A better approach should consider the whole user event telemetry flows.
+
+WITH
+{% if is_incremental() %}
+max_time AS (
```

Comment:
Is this `max_time` or `mt`?

Reply:
`max_time` the name of the CTE. `mt` is the alias in the `JOIN` statement.

Reply:
So there is no need to have a `JOIN make_table mt ...` then?

Reply:
The role of `max_time` CTE is to enable incremental loading in the current table. It provides the timestamp of the latest record from the source table. The value is stored in column `max_time` of the CTE. The trick is that the source table is the union of multiple tables, so the latest timestamp from each of the original source tables is not the same value. For this reason, the latest record's timestamp is stored per each source table (`_dbt_source_relation2`).

When the actual incremental join happens, the source table is filtered by joining the original table with `max_time` CTE. `mt` is the alias of the CTE in this join. 

One thing that can be confusing is that both the CTE and column are called `max_time`. Perhaps the CTE can be renamed to `latest_records` and the column to `latest_received_at`.

Reply:
should we use `timestamp` here, instead of `received_date`, since clustering index is on timestamp.

Reply:
@sri-byte [Snowflake documentation](https://docs.snowflake.com/en/user-guide/tables-clustering-keys.html#strategies-for-selecting-clustering-keys) suggest to use columns that are most actively used in selective filters. Based on the queries provided, it seems that in the `WHERE` part of the queries the timestamp is cast to date. This version of the table builds on that, adding date as a separate column. Queries should be updated to utilize the `event_date` column rather than performing a cast.